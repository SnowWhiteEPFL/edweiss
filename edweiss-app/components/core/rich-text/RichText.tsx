import Colors, { Color } from '@/constants/Colors';
import ReactComponent from '@/constants/Component';
import { MarginProps, PaddingProps, PredefinedSizes, Size, computeMargins, computeSize, computeSizeOpt, paddingSizes } from '@/constants/Sizes';
import useTheme from '@/hooks/theme/useTheme';
import { useOptionalColor } from '@/hooks/theme/useThemeColor';
import React, { useMemo } from 'react';
import { ViewStyle } from 'react-native';
import MathJax from 'react-native-mathjax';
import sanitizeHtml from 'sanitize-html';

type RichTextProps = PaddingProps & MarginProps & {
	backgroundColor?: Color,
	color?: Color,
	children: string,
	size?: Size,
	font?: "Inter" | "Times New Roman"
}

const headerSizes = [
	32,
	28,
	24,
	20,
	18,
	16
]

const richTextSizes: PredefinedSizes = {
	xs: 0.8,
	sm: 0.9,
	md: 1.0,
	lg: 1.2,
	xl: 1.4
}

function headerLevel(line: string, level: number) {
	return `<div style="font-size: ${headerSizes[level - 1]}px; font-weight: bold">${line.substring(level + 1)}</div>`;
}

export function richTextToHTML(rawText: string): string {
	const text = sanitizeHtml(rawText, {
		allowedTags: [] // no tags allowed so no injection attacks are permitted
	}).trim();

	const lines = text.split("\n");

	let res = "";

	main:
	for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
		if (lineIndex != 0) {
			res += '<br/>';
		}

		const line = lines[lineIndex].trim();

		for (let headerCheckIndex = 0; headerCheckIndex < 6; headerCheckIndex++) {
			if (line.startsWith("#".repeat(headerCheckIndex + 1) + " ")) {
				res += headerLevel(line, headerCheckIndex + 1);
				continue main;
			}
		}

		res += line;
	}

	return res;
}

const RichText: ReactComponent<RichTextProps> = ({ size = "md", ...props }) => {
	const text = useMemo(() => richTextToHTML(props.children), [props.children]);

	const theme = useTheme();
	const computedBackgroundColor = useOptionalColor(props.backgroundColor);
	const computedColor = useOptionalColor(props.color);

	return (
		<MathJax
			style={{
				backgroundColor: computedBackgroundColor ?? 'transparent',
				...computeMargins(props)
			} satisfies ViewStyle}
			html={`
				<style>
					@font-face {
						font-family: Inter;
						src: url('file:///android_asset/fonts/Inter.ttf') format('truetype')
					}
					body {
						width: 100%;
						height: 100%;

						margin: 0 !important;
 						padding: ${computeSizeOpt(props.p, paddingSizes) ?? 0} !important;
						padding-top: ${computeSizeOpt(props.pt, paddingSizes) ?? 0} !important;
						padding-bottom: ${computeSizeOpt(props.pb, paddingSizes) ?? 0} !important;
						padding-left: ${computeSizeOpt(props.pl, paddingSizes) ?? 0} !important;
						padding-right: ${computeSizeOpt(props.pr, paddingSizes) ?? 0} !important;
						
						font-family: ${props.font ?? "Inter"};
						color: ${computedColor ?? Colors[theme].text};
						font-size: ${16 * computeSize(size, richTextSizes)}px;
						line-height: 1.6;
					}
				</style>
				<body>
					${text}
				</body>
			`}
			scrollEnabled={false}
			allowBounce={false}
			alwaysBounceHorizontal={false}
			alwaysBounceVertical={false}
			bounces={false}
		/>
	);
};

export default RichText;
