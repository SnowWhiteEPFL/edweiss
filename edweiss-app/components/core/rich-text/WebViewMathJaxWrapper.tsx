import ReactComponent from '@/constants/Component';
import React, { useMemo, useState } from 'react';
import { View, ViewStyle } from 'react-native';
import { WebView, WebViewProps } from 'react-native-webview';

const mathjaxConfigPayload = JSON.stringify({
	messageStyle: 'none',
	extensions: ['tex2jax.js'],
	jax: ['input/TeX', 'output/HTML-CSS'],
	tex2jax: {
		inlineMath: [['$', '$'], ['\\(', '\\)']],
		displayMath: [['$$', '$$'], ['\\[', '\\]']],
		processEscapes: true,
	},
	TeX: {
		extensions: ['AMSmath.js', 'AMSsymbols.js', 'noErrors.js', 'noUndefined.js']
	}
});

interface WebViewJaxWrapperProps {
	source: string,
	style?: ViewStyle,
	webview?: WebViewProps,
	disableMathJax?: boolean
}

const WebViewMathJaxWrapper: ReactComponent<WebViewJaxWrapperProps> = (props) => {
	const [height, setHeight] = useState(0);

	const html = useMemo(() => `
		<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
		${props.disableMathJax ? props.source : `
			<script type="text/x-mathjax-config">
				MathJax.Hub.Config(${mathjaxConfigPayload});
				MathJax.Hub.Queue(() => {
					window.ReactNativeWebView.postMessage(String(document.documentElement.scrollHeight));
					document.getElementById("jax-content").style.visibility = 'visible';
				});
			</script>
			<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js"></script>
			<div id="jax-content" style="visibility: hidden;">${props.source}</div>
		`}
	`, [props.source]);

	return (
		<View style={[{ height }, props.style]}>
			<WebView
				source={{ html }}
				onMessage={message => setHeight(Number(message.nativeEvent.data))}
				cacheEnabled
				cacheMode='LOAD_CACHE_ELSE_NETWORK'
				{...props.webview}
			/>
		</View>
	);
};

export default WebViewMathJaxWrapper;

// const [mathJax, setMathJax] = useState<string | undefined>(cachedMathJax);

// useEffect(() => {
// 	if (!mathJax) {
// 		(async () => {
// 			const res = await fetch("https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js");
// 			const content = await res.text();
// 			setMathJax(content);
// 			cachedMathJax = content;
// 		})();
// 	}
// }, []);

// <script type="text/javascript">${mathJax}</script>
