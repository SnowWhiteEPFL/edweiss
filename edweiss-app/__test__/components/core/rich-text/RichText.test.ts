import { richTextToHTML } from '@/components/core/rich-text/RichText';

jest.mock("react-native-webview", () => ({
	WebView: jest.fn()
}));

describe("RichText", () => {
	it("should parse correctly", () => {
		const res = richTextToHTML(`
			# Title
			## Subtitle
			Hello, $a + b$
		`, 'light');
		expect(res).toBe(`<div style="font-size: 32px; font-weight: bold">Title</div><div style="font-size: 28px; font-weight: bold">Subtitle</div>Hello, $a + b$`);
	});

	it("should parse code and comments correctly", () => {
		const res = richTextToHTML(`
${"```typescript"}
// the comment is well rendered
/*
multiline comments
*/
let a = 15;
class Important {
	export const msg = "hi mom";
}
${"```"}
${"```rust"}
fn why_love_rust() {
	println!("Because Rust is best");
}
${"```"}
		`, 'light');
		expect(res).toMatchSnapshot();
	});
});
