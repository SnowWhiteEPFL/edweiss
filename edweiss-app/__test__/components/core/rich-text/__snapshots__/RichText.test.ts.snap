// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichText should parse code and comments correctly 1`] = `
"
			<div style="
				padding: 10px;
				padding-top: 10px;
				margin: 10px;
				border-radius: 8px;
				position: relative;
				font-family: Fira Code;
				background-color: #dce0e8;
				"><div style="position: absolute; right: 8px; top: 4px; font-size: 14px; color: #9ca0b0">typescript</div>
	<div style="display: flex">
		<div style="color: #bcc0cc; margin-right: 16px">
			<div style="text-align: end">1</div><div style="text-align: end">2</div><div style="text-align: end">3</div><div style="text-align: end">4</div><div style="text-align: end">5</div><div style="text-align: end">6</div><div style="text-align: end">7</div><div style="text-align: end">8</div>
		</div>
		<div style="white-space: pre; overflow: scroll"><span style="color: #9ca0b0">// the comment is well rendered</span><br/><span style="color: #9ca0b0">/*<br/>multiline comments<br/>*/</span><br/><span style="color: #8839ef">let</span> <span style="color: #d20f39">a</span> = <span style="color: #fe640b">15</span>;<br/><span style="color: #8839ef">class</span> <span style="color: #d20f39">Important</span> {<br/>  <span style="color: #d20f39">export</span> <span style="color: #8839ef">const</span> <span style="color: #d20f39">msg</span> = <span style="color: #40a02b">"hi mom"</span>;<br/>}</div>
	</div>
	</div>
			
			<div style="
				padding: 10px;
				padding-top: 10px;
				margin: 10px;
				border-radius: 8px;
				position: relative;
				font-family: Fira Code;
				background-color: #dce0e8;
				"><div style="position: absolute; right: 8px; top: 4px; font-size: 14px; color: #9ca0b0">rust</div>
	<div style="display: flex">
		<div style="color: #bcc0cc; margin-right: 16px">
			<div style="text-align: end">1</div><div style="text-align: end">2</div><div style="text-align: end">3</div>
		</div>
		<div style="white-space: pre; overflow: scroll"><span style="color: #8839ef">fn</span> <span style="color: #1e66f5">why_love_rust</span>() {<br/>  <span style="color: #1e66f5">println!</span>(<span style="color: #40a02b">"Because Rust is best"</span>);<br/>}</div>
	</div>
	</div>
			"
`;
